components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc

```diff
@@ -0,0 +1,384 @@
+    uint32_t new_block_height) {
+  // Query block info by block height
+  auto block_id =
+      zcash::mojom::BlockID::New(new_block_height, std::vector<uint8_t>());
+  zcash_rpc_->GetTreeState(
+      chain_id_, std::move(block_id),
+      base::BindOnce(&ZCashShieldSyncService::OnGetTreeStateForChainReorg,
+                     weak_ptr_factory_.GetWeakPtr(), new_block_height));
+}
+
+void ZCashShieldSyncService::OnGetTreeStateForChainReorg(
+    uint32_t new_block_height,
+    base::expected<zcash::mojom::TreeStatePtr, std::string> tree_state) {
+  if (!tree_state.has_value() || !tree_state.value() ||
+      new_block_height != (*tree_state)->height) {
+    error_ = Error{kFailedToReceiveTreeState, tree_state.error()};
+    ScheduleWorkOnTask();
+    return;
+  } else {
+    // Reorg database so records related to removed blocks are wiped out
+    background_orchard_storage_
+        .AsyncCall(&ZCashOrchardStorage::HandleChainReorg)
+        .WithArgs(account_id_.Clone(), (*tree_state)->height,
+                  (*tree_state)->hash)
+        .Then(base::BindOnce(
+            &ZCashShieldSyncService::OnDatabaseUpdatedForChainReorg,
+            weak_ptr_factory_.GetWeakPtr(), (*tree_state)->height));
+  }
+}
+
+void ZCashShieldSyncService::OnDatabaseUpdatedForChainReorg(
+    uint32_t new_block_height,
+    std::optional<ZCashOrchardStorage::Error> error) {
+  if (error) {
+    error_ = Error{kFailedToUpdateDatabase, error->message};
+    ScheduleWorkOnTask();
+  } else {
+    latest_scanned_block_ = new_block_height;
+    // Reload list of discovered spendable notes and update account info
+    UpdateSpendableNotes();
+  }
+}
+
+void ZCashShieldSyncService::UpdateSpendableNotes() {
+  spendable_notes_ = std::nullopt;
+  background_orchard_storage_.AsyncCall(&ZCashOrchardStorage::GetSpendableNotes)
+      .WithArgs(account_id_.Clone())
+      .Then(base::BindOnce(&ZCashShieldSyncService::OnGetSpendableNotes,
+                           weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::OnGetSpendableNotes(
+    base::expected<std::vector<OrchardNote>, ZCashOrchardStorage::Error>
+        result) {
+  if (!result.has_value()) {
+    error_ = Error{kFailedToRetrieveSpendableNotes, result.error().message};
+  } else {
+    spendable_notes_ = result.value();
+    current_sync_status_ = mojom::ZCashShieldSyncStatus::New(
+        latest_scanned_block_.value(), chain_tip_block_.value(),
+        spendable_notes_->size(), GetSpendableBalance());
+  }
+
+  ScheduleWorkOnTask();
+}
+
+void ZCashShieldSyncService::UpdateChainTip() {
+  zcash_rpc_->GetLatestBlock(
+      chain_id_, base::BindOnce(&ZCashShieldSyncService::OnGetLatestBlock,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::OnGetLatestBlock(
+    base::expected<zcash::mojom::BlockIDPtr, std::string> result) {
+  if (!result.has_value()) {
+    error_ = Error{kFailedToUpdateChainTip, result.error()};
```

Comment:
Should we be validating this value is greater than the current chain tip before updating it?

Reply:
There is a check in the WorkOnTask method
```
  if (*latest_scanned_block_ < *chain_tip_block_) {
    DownloadBlocks();
    return;
  }
```

Reply:
Ahh, thanks for pointing that out. I missed that during my review.